crates/sui-core/src/state_accumulator.rs

```diff
@@ -78,88 +78,144 @@ impl StateAccumulator {
 
     /// Accumulates given effects and returns the accumulator without side effects.
     pub fn accumulate_effects(&self, effects: Vec<TransactionEffects>) -> Accumulator {
+        // prefetch all objects that may need to be looked up
+
+        // get all unwrapped ObjectRefs
+        let all_unwrapped = effects
+            .iter()
+            .flat_map(|fx| {
+                fx.unwrapped()
+                    .iter()
+                    .map(|(oref, _owner)| *oref)
+                    .chain(fx.unwrapped_then_deleted().iter().cloned())
+                    .collect::<Vec<ObjectRef>>()
+            })
+            .collect::<Vec<ObjectRef>>();
+
+        let all_wrapped: Vec<ObjectRef> = all_unwrapped
+            .iter()
+            .filter_map(|(id, seq_num, _digest)| {
+                self.authority_store
+                    .get_object_ref_prior_to_key(id, *seq_num)
+                    .expect("read cannot fail")
+            })
+            .collect();
+
+        let modified_at_version_keys: Vec<ObjectKey> = effects
+            .iter()
+            .flat_map(|fx| {
+                fx.modified_at_versions()
+                    .iter()
+                    .map(|(id, seq_num)| ObjectKey(*id, *seq_num))
+            })
+            .collect();
+
+        let all_modified_at: HashMap<ObjectKey, ObjectRef> = self
+            .authority_store
+            .multi_get_object_by_key(&modified_at_version_keys.clone())
+            .expect("Failed to get modified_at_versions object from object table")
+            .into_iter()
+            .zip(modified_at_version_keys)
+            // it's safe to filter missing objects here, as they could be unwrapped objects
+            // that did not exist previously. If they should be there, but are not, the caller
+            // to this function will catch it
```

Comment:
This is a new comment. And I think so. If the caller expects there to be a mapping in the cache and it isn't, then it is equivalent to when this function does the lookup against the database and doesn't find it. The difference is that the caller has better context on what is ok to not exist and what isn't (for example, when we filter out `unwrapped_ids` in the caller. We can't do that in this function since it's the aggregate of all effects (this would be the bug that @lxfind found)

Reply:
Oh sorry I understand the confusion. I mean to say "callee" here. i.e. the function that we pass the lookup_cache to. Will update